db/migrations/20211123120031_improve_page_views.sql

```diff
@@ -0,0 +1,104 @@
+-- Page views should be able to be queried even if they aren't posts.
+
+-- #############################################################################
+-- migrate:up
+-- #############################################################################
+
+CREATE OR REPLACE FUNCTION get_path_from_url (url TEXT)
+RETURNS TEXT AS $$
+  SELECT (regexp_matches(url, '^https?://[^/]+(.*)'))[1]
+$$ LANGUAGE sql IMMUTABLE STRICT;
+
+CREATE OR REPLACE FUNCTION get_base_path_from_path (path TEXT)
+RETURNS TEXT AS $$
+  SELECT (regexp_matches(path, '^(/|(/[^?^#^/]+)+)/?'))[1]
+$$ LANGUAGE sql IMMUTABLE STRICT;
+
+DROP VIEW page_view;
+
+CREATE VIEW page_view AS (
+  SELECT
+    environment,
+    event_type,
+    timestamp,
+    tab_id,
+    user_id,
+    client_id,
+    post_id,
+    path,
+    base_path
+  FROM (
+    SELECT
+      environment,
+      'page_load_finished' AS event_type,
+      timestamp,
+      tab_id,
+      user_id,
+      client_id,
+      post_id,
+      get_path_from_url(url) AS path,
+      get_base_path_from_path(get_path_from_url(url)) AS base_path
+    FROM event_page_load_finished
+    UNION ALL
+    SELECT
+      environment,
+      'navigate' AS event_type,
+      timestamp,
+      tab_id,
+      user_id,
+      client_id,
+      to_post_id AS post_id,
+      "to" AS path,
+      get_base_path_from_path("to") AS base_path
+    FROM event_navigate
+  ) a
+);
+
+-- JP why do you need the second pair of parenthesis there? A: no idea
+CREATE INDEX IF NOT EXISTS raw_event_type_base_path_navigate ON raw (get_base_path_from_path(event->>'to')) WHERE event_type = 'navigate';
+CREATE INDEX IF NOT EXISTS raw_event_type_base_path_page_load_finished ON raw (get_base_path_from_path(get_path_from_url(event->>'url'))) WHERE event_type = 'pageLoadFinished';
+CREATE INDEX IF NOT EXISTS raw_event_type_path_page_load_finished ON raw (get_path_from_url(event->>'url')) WHERE event_type = 'pageLoadFinished';
+
+-- #############################################################################
+-- migrate:down
+-- #############################################################################
```

Comment:
Why are the `up` and `down` in the same file?

Reply:
That's the way dbmate does it. I agree it's not what you'd expect.